packages/object/src/Charcoal/Object/RevisionsManager.php

```diff
@@ -0,0 +1,227 @@
+    {
+        $model       = $this->getModel();
+        $modelConfig = $this->getModelRevisionConfig($model);
+        $properties  = array_keys($model->data());
+
+        if ($modelConfig->hasProperties()) {
+            return array_intersect($properties, $modelConfig->getProperties());
+        }
+
+        if ($modelConfig->hasExcludedProperties()) {
+            $excludedProperties = $modelConfig->getExcludedProperties();
+
+            if ($modelConfig->hasIncludedProperties()) {
+                $includedProperties = $modelConfig->getIncludedProperties();
+                $excludedProperties = array_filter($excludedProperties, fn($e) => !in_array($e, $includedProperties));
+            }
+
+            return array_filter(
+                $properties,
+                fn($n) => !in_array($n, $excludedProperties)
+            );
+        }
+
+        return $properties;
+    }
+
+    public function getObjectRevisionClass(): string
+    {
+        $modelConfig = $this->getModelRevisionConfig();
+
+        return $modelConfig->getRevisionClass();
+    }
+
+    public function createRevisionObjectCollectionLoader(): CollectionLoader
+    {
+        return new CollectionLoader([
+            'logger'  => $this->logger,
+            'factory' => $this->modelFactory(),
+            'model'   => $this->getRevisionObjectPrototype($this->getObjectRevisionClass()),
+        ]);
+    }
+
+    public function getRevisionObjectPrototype(): ObjectRevisionInterface
+    {
+        return $this->modelFactory()->get($this->getObjectRevisionClass());
+    }
+
+    public function createRevisionObject(): ObjectRevisionInterface
+    {
+        return $this->modelFactory()->create($this->getObjectRevisionClass());
+    }
+
+    public function revisionFromNumber(int $number): ObjectRevisionInterface
+    {
+        return $this->createRevisionObject()->objectRevisionNum($this->getModel(), $number);
+    }
+
+    public function revisionEnabled(): bool
+    {
+        $model          = $this->getModel();
+        $revisionConfig = $this->getModelRevisionConfig($model);
+
+        // If we did not find a config of the value of the config is false, we don't want to revision.
+        if (!$revisionConfig) {
+            return false;
+        }
+
+        return $revisionConfig->isEnabled();
+    }
+
+    private function getModelRevisionConfig(): ?RevisionModelConfig
+    {
+        $model = $this->getModel();
+
+        if (!isset($this->modelRevisionConfig[get_class($model)])) {
+            $this->modelRevisionConfig[get_class($model)] = $this->revisionConfig->buildModelConfig($model);
+        }
+
+        return $this->modelRevisionConfig[get_class($model)];
+    }
+
+    public function getModel(): ModelInterface
+    {
+        if (!isset($this->model)) {
+            throw new \InvalidArgumentException('Setting a `ModelInterface` is imperative to use the revisionManager');
+        }
+
+        return $this->model;
+    }
```

Comment:
With support for PHP parameter/return type-hinting, I wonder if its necessary to still throw exceptions within accessor methods like this. Let PHP throw an error about `getModel()` returning `null`.

Reply:
done